### ISSUE TITLE
Restore grease pencil support

### ISSUE THREAD

**Issue by luboslenco** [armory3d/armory#810]
*But wait for Armory + b2.8 get more mature first.*

**Comment by AdityaSrivastava2819** [comment]
Will we be able to make 2D games too? How about a setting that allows us to lock a particular axis? It would also be better if Armory can render in orthographic mode for 2D. It would allow us to also use 3D objects with 2D Grease Pencil assets. But I am not sure if Grease Pencil is strong enough for making games with... (Grease Pencil is still awesome though.)

**Comment by mundusnine** [comment]
@AdityaSrivastava2819 Nothing says we can't. Armory is a 3D rendered engine but Unity does it's 2D with the 3D renderer a bit like we would if we used Armory for 2D. The only question right now is do we create a 2D renderer with Armory or do we use the 3D renderer.

**Comment by apaza610** [comment]
I really need this grease pencil support, hope it's soon...

**Comment by N8n5h** [comment]
continued from https://github.com/armory3d/armory/issues/1427, 
@luboslenco I noticed in the grease branch you created, you moved code from iron to armory, so what should be done with the remaining code (Data.hx, SceneFormat.hx, Scene.hx, etc) that was in iron, should be moved to armory too?

**Comment by luboslenco** [comment]
Yes, moved it into single armory branch for a reference. For actual implementation I am thinking of triangulating grease pencil objects in the exporter itself and exporting as a mesh. That way there is no added complexity in the engine itself. Once that runs we could move on to handling animation frames.

**Comment by N8n5h** [comment]
>triangulating grease pencil objects in the exporter itself and exporting as a mesh

Delaunay triangulations? Where should the triangulation be done exactly?

**Comment by luboslenco** [comment]
There is a `Convert Grease Pencil` operator in Blender which might be handy but sadly it does not seem sufficient enough yet to handle the whole task.

The triangulation should be done somewhere along the [exporter.py](https://github.com/armory3d/armory/blob/grease/blender/arm/exporter.py#L1739) execution. Some valuable data may also pop up by digging into grease pencil python api.

**Comment by N8n5h** [comment]
Ok, I ported most of the logic of the library I linked to python, I need to test it though, but supposedly a mesh could be written from just a set of points (maybe strokes points). 

One thing I was wondering is how the process should be done since a grease pencil object can consist of multiple non connected strokes, so what should be considered as a "elegible mesh" when exporting?

**Comment by N8n5h** [comment]
#### Update: 
I've been working to get a triangulation done with the external library [cdt2d](https://github.com/mikolalysenko/cdt2d) and finally got some results, the strokes are simple connected edges and meshes are generated from the triangles:

![](https://media.giphy.com/media/JpkdTo7JGfMi36Tf5l/giphy.gif)

I'm still trying to figure out the render priority for the materials so the fills don't overlap and also a way to render the strokes from the loose edges, not sure where to proceed from here @luboslenco.

Manually applied materials with `Two-Sided` enabled:

![image](https://user-images.githubusercontent.com/42382648/67977367-b5ac3800-fbf6-11e9-87a0-d3d5d10c1452.png)

Krom:

![image](https://user-images.githubusercontent.com/42382648/67977378-bcd34600-fbf6-11e9-9225-9fb929e56bc4.png)

----

#### Note: 
While looking at the grease pencil docs I noticed that strokes have a `triangles` attribute, so I decided to check it out but I couldn't make it work in 2.80. 
After testing I gave it a try in 2.81 beta and it worked. The attribute it's a list of the triangles created in the triangulation when enabling fill on a stroke, so with that info available I decided to port the code to 2.81 to compare results.

After processing the monkey grease pencil object:
2.80: 0.5s
2.81: 0.08s

Also I talked with **antonioya** in a chat where blender devs hang out to get some help with the triangles and according to him apparently they are working on converting grease pencil to meshes, so probably in the future the process could be even faster.

**Comment by N8n5h** [comment]
Update: 
![](https://media.giphy.com/media/LNwu9xAVVlY3w41dWB/giphy.gif)
![](https://media.giphy.com/media/LkkTJJkiV7AcqA5p7b/giphy.gif)
I have been updating my code and now it creates materials automatically for the meshes based on the original fills (basically taking the original fill color).

I also went on the route of physical separation for overlapping meshes for being the fastest for me to get results. Now there is a tiny (0.005) separation in the y axis when creating the meshes using an arbitrary value defined by the user (for now i'm using arm_material_index but plan to add a new one maybe, like "z_index") to have them separated. 
The separation value could be an option since is almost unnoticeable but could be even smaller if the camera will not be far enough.
I have to add a way to mirror the meshes so they appear in the other side too as a to-do, also update my 2.81 code too with this changes.

I'm still trying to figure how to introduce the rendering of the strokes, I'm reading the GreasePencilObject.hx code to have an idea but still I'm not sure how it should be done.

Any thoughts, @luboslenco ?

**Comment by N8n5h** [comment]
Ok, I have some free time again, I'm trying to get to rendering the lines, but I'm not sure how to proceed since there are multiple options. The alternatives I am investigating currently:
* Rendering loose edges like Blender does already:

![image](https://user-images.githubusercontent.com/42382648/71308462-bcc41b00-23db-11ea-9145-8df81a8bfb80.png)
Not sure how much of extra complexity this would add to the engine though.
* Adding a new datatype to handle rendering lines.
* Other solutions I didn't think of yet...

@luboslenco

**Comment by N8n5h** [comment]
### Update: 
My progress so far in my little experiment project:
* Added mesh conversion of grease pencil.
* Added support for a new type of object animation that consists on changing the visibility of meshes every frame, called Visibility Animation.
* Added basic support for polylines in Armory.
* Added a new material property: z_index. It allows solving z-fighting of meshes manually.

![](https://media.giphy.com/media/cno9CuA2AtJxbYvq8Y/giphy.gif)

![](https://media.giphy.com/media/ggiBHn87toD5V3s1ct/giphy.gif)

My experiment is a bit of an alien implementation though... duplicate code in a few different places (things that are already done in iron) to avoid changing iron at least in this prototype.

I plan to move to proper implementation land soon. But I need a few things to be changed to have proper inclusion in the engine: 
* First: Add support for the vertex data that polylines uses. The implementation that I use for polylines requires "previous" and "next", and a other info to be sent as attributes (direction, opacity, width) so I was thinking how to add support to it... 
The hardcoded nature of the Geometry and MeshData currently prevent me from using it to get them to be rendered out-of-the-box like normal meshes. So I'm planning to try to de-hardcode those classes and make them more flexible, with arrays. I'm currently on it and I got something working, I need to fully test it to make sure it works. Is that approach ok, @luboslenco ?

* Second, z_index makes visible an issue and that is that too many shader files (config, frag, vertex) are created for duplicate shaders that only change one uniform value: z_index. 
Currently z_index was implemented easily thanks to "bind_constants", but the issue is that those are per material. So for two polylines or two fills that have the "same" material but different z_index value, duplicate materials have to be created for different z_index values. To fix this the idea is to introduce override "bind_constants" or something like that. Those would exist in the object data maybe, and when checking for uniforms in the Uniform class an extra check would be added to override the value with the one in the object data if it exists. This would save a lot of space, something very important on low-end targets. I thought about making it more general instead of just hardcoding z_index because there might be some possibilities for something more like base color or others...

**Comment by luboslenco** [comment]
Look cool! Some attributes in geometry are packed together (position and normal) to preserve memory and I think even the order of attributes matters so it sounds challenging, but if it turns out to not be a big issue then it should be ok. Alternatively adding support for a few custom slots might do the trick.

**Comment by N8n5h** [comment]
@luboslenco my progress so far: [project-geom](https://github.com/N8n5h/iron/tree/project-geom) [(required change in the exporter)](https://github.com/N8n5h/armory/tree/project-geom), I think it's mostly done but for a detail: I used `haxe.ds.Vector` but probably could use the `raw.vertex_arrays` directly... though I read that `haxe.ds.Vector` works better so I'm not sure.

![](https://media2.giphy.com/media/hQEHtaPVqvFeLEKXja/giphy.gif)
I did multiple tests with different combinations (skinning, uvs, tang) (skinning, vcols) etc. and it's working to the extents of the original, and deinterleaved seems to work too.

Should I change it to use raw.vertex_arrays instead of creating a new array or leave it like that? also, should I open a PR now or group it with the whole grease-pencil project later?

**Comment by luboslenco** [comment]
Using raw.vertex_arrays instead of creating new array sounds good (for js/krom, [haxe.ds.Vector](https://github.com/HaxeFoundation/haxe/blob/development/std/haxe/ds/Vector.hx#L29) is just a regular Array). Please open a PR now, can already start taking it in. 🙂

**Comment by AndreaMonzini** [comment]
Hi,
what is the state of Grease Pencil with Armory recently ?
I am experimenting lottie and svg animations for the web and Armory with Grease Pencil support would be great :) 
Thank you !

**Comment by N8n5h** [comment]
No further changes in my progress since, sorry. :sweat:. I haven't touched the code in quite some time, but last I remember, it was "working" but needed a lot of work to be properly implemented.

**Comment by Zireael07** [comment]
Any updates? I am a heavy grease pencil user...
